### DISCUSSION TITLE
set_io_pin_constraint and pin distance

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3355] [General]
@nerriav It looks like minimum pin distance is only availble globally for a macro, not per set_io_pin_constraint?

https://openroad.readthedocs.io/en/latest/main/src/ppl/README.html

We would like pins to use the entire(initially) width/height of the macro edge to minimize contention at the top level where the macro is used. Currently we add extra horizontal distance in the example below to handle routing congestion at the top level. Routing congestion happens even with horizontal routes, because the horizontal routes will intersect a bit with the PDN.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/83012711-df0f-4b41-bdae-d05d3e3fbc8b)

**Comment by maliberty** [comment]
I don't understand what you are asking here.  Are you talking about -corner_avoidance ?

**Reply by oharboe** [reply]
The issue is that there are properties that are global for the macro, as they are arguments to place_pins command, like minimum distance. In some cases it can be desirable to have these properties per set_io_pin_constraint command (per -group).

Example: below, the pins at the left can be a lot more spread out than those on the bottom.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/0695a52a-09f6-4272-85e9-ed0e0649e990)

**Comment by oharboe** [comment]
@nerriav discovered that `-region:*` and `-region:0-100` to is not identical if the regions is 100 wide. The pins are spread evenly when a range is specified (latter) and bunched together more closely if `*` is specified.

@maliberty Is this intentional?

**Reply by maliberty** [reply]
Not to my knowledge. @eder-matheus please comment.

**Reply by eder-matheus** [reply]
@oharboe It is not intended. I've tried it with my unit tests, and the results with `-region:*` and `-region:0-100` are the same. However, it has a small number of pins.

I don't understand what you mean by "when a range is specified **(latter)**".

Are those pins grouped? If so, they should have the same spacing between each other (i.e., the min distance defined in place_pins -min_distance).

Also, are you running random pin placement or regular pin placement? A reproducible test case would be good.

**Reply by oharboe** [reply]
The ball is in my court, I will produce an example. I need to do so using public information.

**Reply by eder-matheus** [reply]
Thanks! Meanwhile, I'm working on adding the min distance option on set_io_pin_constraint. It looks like a nice addition to the command and should not take many cycles to implement.
